File: src/cpp/KalmanFilter.cpp
#include "KalmanFilter.h"
#include <Eigen/Dense>
#include <boost/math/distributions/chi_squared.hpp>

KalmanFilter::KalmanFilter(Eigen::MatrixXd F, Eigen::MatrixXd B, Eigen::MatrixXd H, 
                           Eigen::MatrixXd x, Eigen::MatrixXd P, Eigen::MatrixXd Q, 
                           Eigen::MatrixXd R)
    : F(F), B(B), H(H), x(x), P(P), Q(Q), R(R) {}

void KalmanFilter::update_state_transition_matrix(double dt) {
    int n = F.rows();
    for (int i = 0; i < n / 2; i++) {
        F(i, i + n / 2) = dt;
    }
}

// Predict step: update the state and error covariance
void KalmanFilter::predict(Eigen::MatrixXd u) {
    x = F * x + B * u;
    P = F * P * F.transpose() + Q;
}

// Update step: update the state and error covariance based on the observation
void KalmanFilter::update(Eigen::MatrixXd z, double outlier_threshold) {
    Eigen::MatrixXd y = z - H * x;
    Eigen::MatrixXd S = H * P * H.transpose() + R;

    // Calculate the Mahalanobis distance
    double D = sqrt((y.transpose() * S.inverse() * y)(0, 0));

    // Calculate the threshold based on the chi-squared distribution
    boost::math::chi_squared chi_squared(y.size());
    double threshold = boost::math::quantile(chi_squared, 1 - outlier_threshold);

    // If the Mahalanobis distance is greater than the threshold,
    // consider the observation as an outlier and skip the update step
    if (D > threshold) {
        return;
    }

    Eigen::MatrixXd K = P * H.transpose() * S.inverse();
    x = x + K * y;
    P = (Eigen::MatrixXd::Identity(P.rows(), P.cols()) - K * H) * P;
}

// Get the current state estimate
Eigen::MatrixXd KalmanFilter::get_current_state() {
    return x;
}

File: src/python/kalman_filter.py

import numpy as np
from scipy.stats import chi2
from numpy.linalg import inv


class KalmanFilter:
    def __init__(self, F, B, H, x, P, Q, R):
        self.F = F
        self.B = B
        self.H = H
        self.x = x
        self.P = P
        self.Q = Q
        self.R = R

    # Update the state transition matrix based on the time step dt
    def update_state_transition_matrix(self, dt):
        n = self.F.shape[0]
        for i in range(n // 2):
            self.F[i, i + n // 2] = dt

    # Predict step: update the state and error covariance
    def predict(self, u):
        self.x = self.F @ self.x + self.B @ u
        self.P = self.F @ self.P @ self.F.T + self.Q

    # Update step: update the state and error covariance based on the observation
    def update(self, z, outlier_threshold=0.01):
        y = z - self.H @ self.x
        S = self.H @ self.P @ self.H.T + self.R

        # Calculate the Mahalanobis distance
        D = np.sqrt((y.T @ inv(S) @ y)[0, 0])

        # Calculate the threshold based on the chi-squared distribution
        threshold = chi2.ppf(1 - outlier_threshold, df=y.shape[0])

        # If the Mahalanobis distance is greater than the threshold,
        # consider the observation as an outlier and skip the update step
        if D > threshold:
            return

        K = self.P @ self.H.T @ inv(S)
        self.x = self.x + K @ y
        self.P = (np.eye(self.P.shape[0]) - K @ self.H) @ self.P

    # Get the current state estimate
    def get_current_state(self):
        return self.x
